### PR TITLE
Update wizard UI and improve e2e tests

### LINKED ISSUE #4972: Wizard: Improve layout of Done screen
## Feature description

The layout of the final screen can be improved a bit by reducing the second border around the iframe and moving the heading over the iframe to instead be a caption:

Before | After
-------|-----
image | image

Also, let's change “Visit in Customizer” to be just “Customize“.

The “Browse AMP Version” could be changed to just “Browse” as well.

It's likely that these buttons will not fit on the same line when translated.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### LINKED ISSUE #4974: Wizard: Improve functionality of Close button vs Finish button
## Feature description

When on the Done screen, the Close button and the Finish button do the same thing:

> image

Should the close button be removed on the Done screen then? There's no sense to “Exit” the wizard when on the Done screen as the icon implies.

Also, I think the Close button should intentionally do `history.back()` to take the user to the screen they were just on (or else pass the return URL via a `return` query param similar to how the Customizer implements the Close button there).

Originally raised in https://github.com/ampproject/amp-wp/pull/4954#issuecomment-652716506.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### LINKED ISSUE #4975: Wizard: Improve template mode selection and Reader theme selection when active theme is also a Reader theme
## Feature description

When a site has a Reader theme already set as the active theme (e.g. Twenty Twenty), when the user expresses they are non-technical then _both_ the Reader mode and the Transitional mode should show as ✅  recommended options.

Additionally, when selecting Reader mode, the list of themes should no longer omit the active theme from the list. Instead, if the user selects the active theme to be the Reader theme, then the template mode should be automatically switched from `reader` to `transitional`, and a notice can appear on the summary screen to make them aware of this. Saying something like:

> Since you selected a Reader theme which is the same as your site's active theme, you've automatically been switched to Transitional template mode.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by johnwatkins0**
## Summary

Fixes #4974 
Fixes #4972 
Fixes #4975 

While working on #4704 I found my local e2e tests were failing intermittently. Rather than saying 🤷 and accepting that they just fail sometimes, I wanted to harden them so they are more predictable and can be a more powerful tool, especially for testing bugs and specific user scenarios, going forward. The e2e tests are in Jest but can cover much more than just the React parts. They could cover everything in the plugin that has a UI and can be described with step-by-step commands to Puppeteer.

### Changes

I used these three small issues as tests cases to improve the e2e tests.

1. Addressed #4974 in c0907e6c9d294fa85b1f15d16c4fe061b57fca34
2. Refactored e2e tests in the next several commits. I think the main thing that was causing unpredictability in the tests was the fact that the wizard's changes persist in the test env DB, so I added a function to reset the environment after the "Done" screen tests. We can skip this function when writing tests that cover changes made on subsequent visits to the wizard. I also needed to make sure we're `await`ing pretty much everything.
3. Addressed #4972 01c76ec393a0889d76d39329fd850b9cc1beae69
4. Addressed #4975 in 52f742c44eac35e50405120b106508ad7daade8f. In this case I wrote the e2e tests first and coded to the tests, which was helpful for covering the scenarios outlined in that issue.

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

**Comment by johnwatkins0** [comment]
Opening this up for review. Reach out if you have any questions. I'm branched off this for #4704.

File: assets/src/setup/components/options-context-provider.js
Comment (johnwatkins0) [inline assets/src/setup/components/options-context-provider.js:50]: I just moved this up because I keep expecting the `fetch` effect to be above the `save` effect.

File: tests/e2e/config/bootstrap.js
Comment (johnwatkins0) [inline tests/e2e/config/bootstrap.js:81]: The previous version of this function was copied from an earlier version of Gutenberg, and the changes here are just a copy+paste of the latest version from Gutenberg.

File: assets/src/setup/components/reader-themes-context-provider.js
Comment (pierlon) [inline assets/src/setup/components/reader-themes-context-provider.js]:
```js
'standard' === themeSupport
```

Why only bail if it's standard mode? Shouldn't it be `'reader' !== themeSupport`, like what was before?

File: assets/src/setup/components/user-context-provider.js
Comment (pierlon) [inline assets/src/setup/components/user-context-provider.js:51]: Why wait to fetch plugin options first before user options? It would be 2 separate requests and they are not mutually exclusive, no?

File: assets/src/setup/index.js
Comment (pierlon) [inline assets/src/setup/index.js]: Maybe it's time to put each on a new line:

```suggestion
import {
	APP_ROOT_ID,
	CURRENT_THEME,
	EXIT_LINK,
	OPTIONS_REST_ENDPOINT,
	READER_THEMES_REST_ENDPOINT,
	UPDATES_NONCE,
	USER_FIELD_DEVELOPER_TOOLS_ENABLED,
	USER_REST_ENDPOINT,
} from 'amp-setup'; // From WP inline script.
```

File: assets/src/setup/pages/save/index.js
Comment (pierlon) [inline assets/src/setup/pages/save/index.js]: I'd recommend reverting this change. When viewed on the page, the term "Visit" seems out of context and persons may not understand what they are "visiting".

image

File: assets/src/setup/pages/save/index.js
Comment (pierlon) [inline assets/src/setup/pages/save/index.js]: Same reasoning as above. Maybe it could instead be something like "Customize AMP page".

File: assets/src/setup/pages/save/index.js
Comment (pierlon) [inline assets/src/setup/pages/save/index.js]: Functionally, this is the same as the "Visit" button, so I suppose they could use the same text.

File: assets/src/setup/pages/template-mode/get-selection-details.js
Comment (pierlon) [inline assets/src/setup/pages/template-mode/get-selection-details.js]: @westonruter should we update the description for transitional mode if the theme is AMP compatiable? Currently, it recommends both transitional and reader mode, but does not inlcude the fact that the user's current theme is AMP compatible:

image

File: assets/src/setup/pages/template-mode/get-selection-details.js
Comment (pierlon) [inline assets/src/setup/pages/template-mode/get-selection-details.js]: I think `currentThemeIsReaderTheme` can be renamed to `currentThemeisAmpCompatible`, since that's what is actually being determined here.

cc @westonruter

File: assets/src/setup/pages/template-mode/screen-ui.js
Comment (pierlon) [inline assets/src/setup/pages/template-mode/screen-ui.js]:
```suggestion
 * @param {Function} props.setCurrentMode The callback to update the selected mode.
```

**Review by pierlon — CHANGES_REQUESTED** [review]
Well done with the refactoring of the E2E tests :clap:.

A few issues I wish to be addressed before merging:

File: assets/src/setup/pages/save/index.js
Comment (westonruter) [reply assets/src/setup/pages/save/index.js]: Maybe it would be more clear if the button were placed _below_ the iframe? There could be a “Browse” and a “Customize” button right under the preview. This could be displayed _instead_ of the “Live view of a page on your site…” text, as the buttons should give that context. 

I was originally thinking that the Dashboard had a “Customize” button but it actually says “Customize Your Site”:

image

Nevertheless, if the buttons were under the iframe then this should make this clear.

Also, the buttons would be placed near the Previous/Finish buttons which keep the CTAs in the same area.

File: assets/src/setup/pages/save/index.js
Comment (westonruter) [reply assets/src/setup/pages/save/index.js]: If not just “Browse” and “Customize” per my comment above, what about “Browse AMP” and “Customize AMP”?

Comment (pierlon) [reply assets/src/setup/pages/save/index.js]: Yea that works as well.

File: assets/src/setup/pages/template-mode/get-selection-details.js
Comment (westonruter) [reply assets/src/setup/pages/template-mode/get-selection-details.js]: Yes, that's a good point. We should export to JS this logic from PHP:

```php
$has_amp_support = current_theme_supports( 'amp' ) || in_array( get_stylesheet(), AMP_Core_Theme_Sanitizer::get_supported_themes(), true );
```

Note I'm intentionally using `get_stylesheet()` here as opposed to `get_template()` because child themes of core themes are often not AMP-compatible.

Nevertheless, we may want to hold off on doing this until I finish my work on the Reader theme loading because this changes how the `amp` theme support is mutated (or not).

There's also a case to do something similar for Reader mode, for themes that _don't_ have `amp` theme support and yet have an `amp` directory, which is captured by this logic at the moment:

https://github.com/ampproject/amp-wp/blob/a8a4d4d05f9727aad2304db103a3ab97817aafa6/includes/class-amp-theme-support.php#L349-L366

We may want to mention that the active theme has special legacy Reader mode support, and also indicate that in the list of Reader themes.

But I suggest filing this in a new issue.

File: assets/src/setup/pages/save/index.js
Comment (pierlon) [reply assets/src/setup/pages/save/index.js]: I like that approach. It would look something like this then:

image

File: assets/src/setup/pages/template-mode/get-selection-details.js
Comment (westonruter) [reply assets/src/setup/pages/template-mode/get-selection-details.js]: It's actually both: current theme is AMP-compatible _and_ it's in the list of Reader themes.

So the name I think is OK. It could also be: `currentThemeIsAmongReaderThemes`.

File: assets/src/setup/components/user-context-provider.js
Comment (johnwatkins0) [reply assets/src/setup/components/user-context-provider.js:51]: This sort of does need to run after options are fetched (when `fetchingOptions` is false) because we need to initialize the value based on `wizard_completed`. If `wizard_completed` is false, then we set the dev tools option to null (requiring the user to make a choice) and in this case we don't need to fetch from the user endpoint at all.

File: assets/src/setup/pages/save/index.js
Comment (johnwatkins0) [reply assets/src/setup/pages/save/index.js]: Updated in b2661a832

File: assets/src/setup/pages/save/index.js
Comment (johnwatkins0) [reply assets/src/setup/pages/save/index.js]: Updated in b2661a832

File: assets/src/setup/pages/template-mode/get-selection-details.js
Comment (johnwatkins0) [reply assets/src/setup/pages/template-mode/get-selection-details.js]: Updated to `currentThemeIsAmongReaderThemes` in 1673aa0e2 because that is more accurate.

File: assets/src/setup/pages/template-mode/get-selection-details.js
Comment (johnwatkins0) [reply assets/src/setup/pages/template-mode/get-selection-details.js]: https://github.com/ampproject/amp-wp/issues/4990

**Comment by johnwatkins0** [comment]
Thanks, @pierlon . I have updated and re-requesting review.

**Comment by westonruter** [comment]
When trying to click on the Reader mode, I noticed there was an invisible element that was preventing my clicks:

image

I've hacked it to hide that in d7dd247b6fa168d46175307aade0771d3641b4f4.

**Comment by westonruter** [comment]
@johnwatkins0 I noticed something unexpected. I went through the Wizard the first time, selected Reader mode, the Twenty Nineteen theme as the Reader theme, and then the mobile redirection toggle was enabled:

> image

However, once I went to the admin settings screen, the toggle was not enabled:

> image

The `POST` request does not have `mobile_redirect` present:

> image

**Comment by westonruter** [comment]
If I toggle the mobile redirect off and then back on, then it gets sent:

> image

> image

**Comment by westonruter** [comment]
@johnwatkins0 One aspect of #4974 was not addressed:

> Also, I think the Close button should intentionally do `history.back()` to take the user to the screen they were just on (or else pass the return URL via a `return` query param similar to how the Customizer implements the Close button there).

Should the `href` of the Close link be populated by `wp_get_referer()` if it doesn't return `false`? If it does return `false`, then it can be set to `menu_page_url( AMP_Options_Manager::OPTION_NAME, false )`, which is what the Finish link should go to. So instead of `exitLink` I think `SetupWizard` should take two props: `closeLink` and `finishLink`. The former should use `wp_get_referrer()` if available (or the admin screen if not), and the latter should always go to the admin setting screen.

**Comment by westonruter** [comment]
In regards to #4975, should the recommendations not be given to users who have DevTools enabled as well? If I have an AMP-compatible theme active, then 

Non-Technical User | Technical User
--------------------|----------------
Screen Shot 2020-07-05 at 22 26 29 | Screen Shot 2020-07-05 at 22 26 39

If something is recommended for a non-technical user, then it _definitely_ should be recommended for a technical user, I should think. Once we have the plugin scan (#4719) implemented then we'll want to recommend Standard mode to Technical users instead when an AMP-compatible theme is being used (either since they have known AMP-compatible theme active, or we determine it is AMP-compatible from a theme scan via #4795).

**Comment by johnwatkins0** [comment]
> @johnwatkins0 I noticed something unexpected. I went through the Wizard the first time, selected Reader mode, the Twenty Nineteen theme as the Reader theme, and then the mobile redirection toggle was enabled:

Fixed in https://github.com/ampproject/amp-wp/pull/4980/commits/7f997eb248790dc39b1d6ec0e9f8367e90eb9af8

**Comment by johnwatkins0** [comment]
> @johnwatkins0 One aspect of #4974 was not addressed:
> 
> > Also, I think the Close button should intentionally do `history.back()` to take the user to the screen they were just on (or else pass the return URL via a `return` query param similar to how the Customizer implements the Close button there).
> 
> Should the `href` of the Close link be populated by `wp_get_referer()` if it doesn't return `false`? If it does return `false`, then it can be set to `menu_page_url( AMP_Options_Manager::OPTION_NAME, false )`, which is what the Finish link should go to. So instead of `exitLink` I think `SetupWizard` should take two props: `closeLink` and `finishLink`. The former should use `wp_get_referrer()` if available (or the admin screen if not), and the latter should always go to the admin setting screen.

Done in 436a87df4

**Comment by johnwatkins0** [comment]
> In regards to #4975, should the recommendations not be given to users who have DevTools enabled as well? If I have an AMP-compatible theme active, then
> 
> Non-Technical User	Technical User
> Screen Shot 2020-07-05 at 22 26 29	Screen Shot 2020-07-05 at 22 26 39
> If something is recommended for a non-technical user, then it _definitely_ should be recommended for a technical user, I should think. Once we have the plugin scan (#4719) implemented then we'll want to recommend Standard mode to Technical users instead when an AMP-compatible theme is being used (either since they have known AMP-compatible theme active, or we determine it is AMP-compatible from a theme scan via #4795).

I made this update in https://github.com/ampproject/amp-wp/pull/4980/commits/fdfa19ba5d15d2f13a292b721a06e618b2af05cb, with the result that:

**User is technical and active theme is reader theme**: Recommend reader and transitional
**User is nontechnical and active theme is reader theme**: Recommend reader and transitional
**User is technical and active theme is not reader theme**: Recommend all equally
**User is nontechnical and active theme is not reader theme**: Recommend reader

All of this will of course be updated for 1.7.

**Review by westonruter — APPROVED** [review]
Functionality checks out! Leaving for @pierlon to approve technical changes.

File: assets/src/setup/components/options-context-provider.js
Comment (pierlon) [inline assets/src/setup/components/options-context-provider.js]: I think it would be better to set it to the original value:

```suggestion
				updatesToSave.mobile_redirect = originalOptions.mobile_redirect;
```

Or what would be better is to merge the default options with the updated ones here, so that no other values are missed. Something like:

```js
updatesToSave = { ...originalOptions, ...updatesToSave };
```

**Review by pierlon — CHANGES_REQUESTED** [review]

File: assets/src/setup/components/options-context-provider.js
Comment (johnwatkins0) [reply assets/src/setup/components/options-context-provider.js]: I think the suggested change works best, because there's no need to send back _all_ the options, just the changed ones.

**Review by pierlon — APPROVED** [review]

**Pull request merged**
